File: mobile/modules/tile-server/ios/PMTilesReader.swift
/**
 * PMTiles v3 reader — header parse + Hilbert tile-ID lookup + byte-range reads.
 *
 * File layout:
 *   [0..127)       Header
 *   [rootDirOffset..+rootDirLength)    Root directory (gzip-compressed)
 *   [leafDirsOffset..+leafDirsLength)  Leaf directories (gzip-compressed)
 *   [tileDataOffset..+tileDataLength)  Tile data blobs
 *
 * Directory entries use LEB128 delta encoding (4 arrays: tileId, runLength,
 * length, offset — each delta-compressed separately).
 */
import Foundation

// PMTiles v3 header (fields we need)
struct PMHeader {
    let rootDirOffset:   UInt64
    let rootDirLength:   UInt64
    let leafDirsOffset:  UInt64
    let tileDataOffset:  UInt64
    let internalComp:    UInt8   // 2 = gzip
}

struct DirEntry {
    let tileId:    UInt64
    let offset:    UInt64
    let length:    UInt32
    let runLength: UInt32  // 0 = leaf directory pointer
}

final class PMTilesReader {
    private let fh:          FileHandle
    private let header:      PMHeader
    private let rootEntries: [DirEntry]
    private var leafCache    = [UInt64: [DirEntry]]()  // offset → parsed leaf
    private let q            = DispatchQueue(label: "tile-server.pmtiles", qos: .userInitiated)

    init(path: String) throws {
        guard let handle = FileHandle(forReadingAtPath: path) else {
            throw NSError(domain: "PMTiles", code: 1,
                          userInfo: [NSLocalizedDescriptionKey: "Cannot open \(path)"])
        }
        fh = handle

        // Read 127-byte header
        fh.seek(toFileOffset: 0)
        let hd = fh.readData(ofLength: 127)
        guard hd.count == 127,
              hd[0] == 0x50, hd[1] == 0x4d, hd[2] == 0x54 else {    // "PMT"
            throw NSError(domain: "PMTiles", code: 2,
                          userInfo: [NSLocalizedDescriptionKey: "Bad magic bytes"])
        }

        func u64(_ o: Int) -> UInt64 {
            hd.withUnsafeBytes { $0.load(fromByteOffset: o, as: UInt64.self).littleEndian }
        }
        header = PMHeader(
            rootDirOffset:  u64(8),
            rootDirLength:  u64(16),
            leafDirsOffset: u64(40),
            tileDataOffset: u64(56),
            internalComp:   hd[97]
        )

        // Parse root directory
        fh.seek(toFileOffset: header.rootDirOffset)
        var raw = fh.readData(ofLength: Int(header.rootDirLength))
        if header.internalComp == 2, let d = gzipDecompress(raw) { raw = d }
        rootEntries = PMTilesReader.parseDir(raw)
    }

    deinit { fh.closeFile() }

    // ── Public: fetch tile bytes ────────────────────────────────────────────
    func tile(z: Int, x: Int, y: Int) -> Data? {
        let id = PMTilesReader.tileId(z: z, x: x, y: y)
        return q.sync { lookup(entries: rootEntries, id: id) }
    }

    // ── Private ─────────────────────────────────────────────────────────────
    private func lookup(entries: [DirEntry], id: UInt64) -> Data? {
        guard let e = bsearch(entries, id) else { return nil }

        if e.runLength == 0 {
            // Leaf directory — cache by its offset key
            let leafEntries: [DirEntry]
            if let cached = leafCache[e.offset] {
                leafEntries = cached
            } else {
                fh.seek(toFileOffset: header.leafDirsOffset + e.offset)
                var raw = fh.readData(ofLength: Int(e.length))
                if header.internalComp == 2, let d = gzipDecompress(raw) { raw = d }
                leafEntries = PMTilesReader.parseDir(raw)
                leafCache[e.offset] = leafEntries
            }
            return lookup(entries: leafEntries, id: id)
        } else {
            fh.seek(toFileOffset: header.tileDataOffset + e.offset)
            return fh.readData(ofLength: Int(e.length))
        }
    }

    private func bsearch(_ entries: [DirEntry], _ id: UInt64) -> DirEntry? {
        var lo = 0, hi = entries.count - 1
        while lo <= hi {
            let mid = (lo + hi) >> 1
            let e   = entries[mid]
            if e.tileId == id { return e }
            if e.tileId  < id { lo = mid + 1 } else { hi = mid - 1 }
        }
        if lo > 0 {
            let prev = entries[lo - 1]
            if prev.runLength == 0 { return prev }                           // leaf pointer
            if id < prev.tileId + UInt64(prev.runLength) { return prev }    // tile run
        }
        return nil
    }

    // ── Directory parsing (LEB128 delta-encoded) ─────────────────────────────
    static func parseDir(_ data: Data) -> [DirEntry] {
        var pos = 0

        func varint() -> UInt64 {
            var r: UInt64 = 0, s = 0
            while pos < data.count {
                let b = UInt64(data[pos]); pos += 1
                r |= (b & 0x7F) << s
                if b & 0x80 == 0 { break }
                s += 7
            }
            return r
        }

        let n = Int(varint())
        var entries = [DirEntry](repeating: DirEntry(tileId:0, offset:0, length:0, runLength:0), count: n)

        // 1. tile IDs (cumulative delta)
        var last: UInt64 = 0
        for i in 0..<n { entries[i] = DirEntry(tileId: last + varint(), offset: 0, length: 0, runLength: 0); last = entries[i].tileId }
        // 2. run lengths
        var rl = [UInt32](repeating: 0, count: n)
        for i in 0..<n { rl[i] = UInt32(varint()) }
        // 3. lengths
        var le = [UInt32](repeating: 0, count: n)
        for i in 0..<n { le[i] = UInt32(varint()) }
        // 4. offsets — PMTiles v3 encodes as (offset+1), with 0 meaning "delta from prev"
        var offsets = [UInt64](repeating: 0, count: n)
        for i in 0..<n {
            let tmp = varint()
            if i > 0 && tmp == 0 {
                offsets[i] = offsets[i-1] + UInt64(le[i-1])
            } else {
                offsets[i] = tmp > 0 ? tmp - 1 : 0
            }
        }

        for i in 0..<n {
            entries[i] = DirEntry(tileId: entries[i].tileId, offset: offsets[i], length: le[i], runLength: rl[i])
        }
        return entries
    }

    // ── Hilbert curve: (z, x, y) → PMTiles tile ID ──────────────────────────
    static func tileId(z: Int, x: Int, y: Int) -> UInt64 {
        if z == 0 { return 0 }
        // Base = Σ 4^i for i in [0, z-1] = (4^z - 1) / 3
        let base: UInt64 = (UInt64(1) << (2 * z) &- 1) / 3
        let n = 1 << z
        var x = x, y = y, d: UInt64 = 0, s = n >> 1
        while s > 0 {
            let rx = (x & s) != 0 ? 1 : 0
            let ry = (y & s) != 0 ? 1 : 0
            d += UInt64(s * s) * UInt64((3 * rx) ^ ry)
            if ry == 0 {
                if rx == 1 { x = s - 1 - x; y = s - 1 - y }
                let t = x; x = y; y = t
            }
            s >>= 1
        }
        return base &+ d
    }
}

File: mobile/modules/tile-server/android/src/main/java/expo/modules/tileserver/PMTilesReader.kt
package expo.modules.tileserver

import java.io.ByteArrayOutputStream
import java.io.RandomAccessFile
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.zip.GZIPInputStream

/**
 * PMTiles v3 reader — identical logic to the iOS Swift version.
 * Uses RandomAccessFile for thread-safe byte-range reads after construction.
 * Leaf directory entries are cached to avoid re-parsing on repeated access.
 */
data class DirEntry(val tileId: Long, val offset: Long, val length: Int, val runLength: Int)

data class PMHeader(
    val rootDirOffset:  Long,
    val rootDirLength:  Long,
    val leafDirsOffset: Long,
    val tileDataOffset: Long,
    val internalComp:   Int
)

class PMTilesReader(path: String) : AutoCloseable {
    private val raf        = RandomAccessFile(path, "r")
    private val header: PMHeader
    private val rootEntries: List<DirEntry>
    private val leafCache  = HashMap<Long, List<DirEntry>>()

    init {
        val hd = ByteArray(127).also { raf.seek(0); raf.readFully(it) }
        require(hd[0] == 'P'.code.toByte() && hd[1] == 'M'.code.toByte()) { "Bad PMTiles magic" }
        fun u64(o: Int) = ByteBuffer.wrap(hd, o, 8).order(ByteOrder.LITTLE_ENDIAN).long
        header = PMHeader(
            rootDirOffset  = u64(8),
            rootDirLength  = u64(16),
            leafDirsOffset = u64(40),
            tileDataOffset = u64(56),
            internalComp   = hd[97].toInt() and 0xFF
        )
        rootEntries = parseDir(readInternalBytes(header.rootDirOffset, header.rootDirLength.toInt()))
    }

    override fun close() = raf.close()

    // ── Public ────────────────────────────────────────────────────────────────
    @Synchronized
    fun tile(z: Int, x: Int, y: Int): ByteArray? {
        val id = tileId(z, x, y)
        return lookup(rootEntries, id)
    }

    // ── Private ───────────────────────────────────────────────────────────────
    private fun lookup(entries: List<DirEntry>, id: Long): ByteArray? {
        val e = bsearch(entries, id) ?: return null
        return if (e.runLength == 0) {
            // Leaf directory
            val leaf = leafCache.getOrPut(e.offset) {
                parseDir(readInternalBytes(header.leafDirsOffset + e.offset, e.length))
            }
            lookup(leaf, id)
        } else {
            readRawBytes(header.tileDataOffset + e.offset, e.length)
        }
    }

    private fun readRawBytes(offset: Long, length: Int): ByteArray =
        ByteArray(length).also { raf.seek(offset); raf.readFully(it) }

    private fun readInternalBytes(offset: Long, length: Int): ByteArray {
        val raw = ByteArray(length).also { raf.seek(offset); raf.readFully(it) }
        return if (header.internalComp == 2) gunzip(raw) ?: raw else raw
    }

    private fun bsearch(entries: List<DirEntry>, id: Long): DirEntry? {
        var lo = 0; var hi = entries.size - 1
        while (lo <= hi) {
            val mid = (lo + hi) ushr 1
            val e   = entries[mid]
            when {
                e.tileId == id -> return e
                e.tileId  < id -> lo = mid + 1
                else            -> hi = mid - 1
            }
        }
        if (lo > 0) {
            val prev = entries[lo - 1]
            if (prev.runLength == 0) return prev
            if (id < prev.tileId + prev.runLength) return prev
        }
        return null
    }

    // ── Directory parsing (LEB128 delta-encoded) ─────────────────────────────
    companion object {
        fun parseDir(data: ByteArray): List<DirEntry> {
            var pos = 0
            fun varint(): Long {
                var r = 0L; var s = 0
                while (pos < data.size) {
                    val b = data[pos++].toLong() and 0xFF
                    r = r or ((b and 0x7F) shl s)
                    if (b and 0x80 == 0L) break
                    s += 7
                }
                return r
            }
            val n     = varint().toInt()
            val ids   = LongArray(n); var last = 0L
            for (i in 0 until n) { ids[i] = last + varint(); last = ids[i] }
            val rls   = IntArray(n)  { varint().toInt() }
            val lens  = IntArray(n)  { varint().toInt() }
            val offs  = LongArray(n)
            for (i in 0 until n) {
                val tmp = varint()
                offs[i] = if (i > 0 && tmp == 0L) {
                    offs[i - 1] + lens[i - 1]
                } else if (tmp > 0) {
                    tmp - 1
                } else {
                    0
                }
            }
            return List(n) { DirEntry(ids[it], offs[it], lens[it], rls[it]) }
        }

        fun tileId(z: Int, x: Int, y: Int): Long {
            if (z == 0) return 0
            val base = ((1L shl (2 * z)) - 1) / 3
            val n    = 1 shl z
            var cx = x; var cy = y; var d = 0L; var s = n shr 1
            while (s > 0) {
                val rx = if ((cx and s) != 0) 1 else 0
                val ry = if ((cy and s) != 0) 1 else 0
                d += s.toLong() * s.toLong() * ((3 * rx) xor ry).toLong()
                if (ry == 0) {
                    if (rx == 1) { cx = s - 1 - cx; cy = s - 1 - cy }
                    val t = cx; cx = cy; cy = t
                }
                s = s shr 1
            }
            return base + d
        }

        fun gunzip(data: ByteArray): ByteArray? = try {
            val out = ByteArrayOutputStream()
            GZIPInputStream(data.inputStream()).use { it.copyTo(out) }
            out.toByteArray()
        } catch (e: Exception) { null }
    }
}
